src/util-streaming-buffer.c

```diff
@@ -0,0 +1,689 @@
+    if (ptr != NULL) {
+        /* for safe printing and general caution, lets memset the
+         * new data to 0 */
+        size_t diff = grow - sb->buf_size;
+        void *new_mem = ((char *)ptr) + sb->buf_size;
+        memset(new_mem, 0, diff);
+
+        sb->buf = ptr;
+        sb->buf_size = grow;
+        SCLogDebug("grown buffer to %u", grow);
+    }
+}
+
+static void Grow(StreamingBuffer *sb)
+{
+    uint32_t grow = sb->buf_size * 2;
+    void *ptr = REALLOC(sb->cfg, sb->buf, sb->buf_size, grow);
+    if (ptr != NULL) {
+        /* for safe printing and general caution, lets memset the
+         * new data to 0 */
+        size_t diff = grow - sb->buf_size;
+        void *new_mem = ((char *)ptr) + sb->buf_size;
+        memset(new_mem, 0, diff);
+
+        sb->buf = ptr;
+        sb->buf_size = grow;
+        SCLogDebug("grown buffer to %u", grow);
+    }
+}
+
+/**
+ *  \brief slide to absolute offset
+ *  \todo if sliding beyond window, we could perhaps reset?
+ */
+void StreamingBufferSlideToOffset(StreamingBuffer *sb, uint64_t offset)
+{
+    if (offset > sb->stream_offset &&
+        offset <= sb->stream_offset + sb->buf_offset)
+    {
+        uint32_t slide = offset - sb->stream_offset;
+        uint32_t size = sb->buf_offset - slide;
+        SCLogDebug("sliding %u forward, size of original buffer left after slide %u", slide, size);
+        memmove(sb->buf, sb->buf+slide, size);
+        sb->stream_offset += slide;
+        sb->buf_offset = size;
+    }
+}
+
+void StreamingBufferSlide(StreamingBuffer *sb, uint32_t slide)
+{
+    uint32_t size = sb->buf_offset - slide;
+    SCLogDebug("sliding %u forward, size of original buffer left after slide %u", slide, size);
+    memmove(sb->buf, sb->buf+slide, size);
+    sb->stream_offset += slide;
+    sb->buf_offset = size;
+}
+
+#define DATA_FITS(sb, len) \
+    ((sb)->buf_offset + (len) <= (sb)->buf_size)
+
+StreamingBufferSegment *StreamingBufferAppendRaw(StreamingBuffer *sb, const uint8_t *data, uint32_t data_len)
+{
+    StreamingBufferSegment *seg = CALLOC(sb->cfg, 1, sizeof(StreamingBufferSegment));
+    if (seg != NULL) {
+        if (!DATA_FITS(sb, data_len)) {
+            if (sb->cfg->flags & STREAMING_BUFFER_AUTOSLIDE)
+                AutoSlide(sb);
+            if (sb->buf_size == 0) {
+                GrowToSize(sb, data_len);
+            } else {
+                while (!DATA_FITS(sb, data_len)) {
+                    Grow(sb);
+                }
+            }
+        }
+        if (!DATA_FITS(sb, data_len)) {
+            FREE(sb->cfg, seg, sizeof(StreamingBufferSegment));
+            return NULL;
+        }
+
+        memcpy(sb->buf + sb->buf_offset, data, data_len);
+        seg->stream_offset = sb->stream_offset + sb->buf_offset;
+        seg->segment_len = data_len;
+        sb->buf_offset += data_len;
+        return seg;
+    }
+    return NULL;
+}
+
+void StreamingBufferAppend(StreamingBuffer *sb, StreamingBufferSegment *seg,
```

Comment:
Maybe this should be StreamingBufferAppendSegment? While the one below can be just StreamBufferAppend instead of NoTrack - better describes the function and the arguments required.

Regarding my comment above on documentation: What do you see as the common use case? With or without segments.  For the DNP3 buffers, I'd probably not use the segments, unless there was a compelling reason to do so?